### PR TITLE
Remove farm.get_pool_addresses(), avoid race condition in Poolmanager

### PR DESCRIPTION
**Pull request by StefanOberhumer**

### CODE REVIEW

File: libpoolprotocols/PoolManager.cpp [hunk 1]
```diff
@@ -319,7 +316,7 @@ void PoolManager::setActiveConnection(unsigned int idx)
     }
 }
 
-const URI *PoolManager::getActiveConnection()
+const URI* PoolManager::getActiveConnection()
 {
     if (m_connections.size() > m_activeConnectionIdx)
         return &m_connections.at(m_activeConnectionIdx);
```
Comment (jean-m-cyr) [inline libpoolprotocols/PoolManager.cpp]: More like C++ would be:

`return &m_connections[m_activeConnectionIdx];`

Comment (jean-m-cyr) [reply libpoolprotocols/PoolManager.cpp]: Or, simply:

`return m_connections.at + m_activeConnectionIdx;`

Comment (AndreaLanfranchi) [reply libpoolprotocols/PoolManager.cpp]: I'm always puzzled about direct random access on vectors. Any index out of bounds does not throw while .at() method does.

Personally I'd simplify the with the single line

`return &m_connections.at(m_activeConnectionIdx);`

as m_activeConnectionIdx is checked to be inbounds every time is set.

Comment (jean-m-cyr) [reply libpoolprotocols/PoolManager.cpp]: That is precisely why I dislike the .at() notation. The out-of-bounds case is already taken care of by the preceding if, and I prefer the implied understanding of pointer arithmetic of using the perfectly equivalent:

`return m_connections + m_activeConnectionIdx;`

Comment (jean-m-cyr) [reply libpoolprotocols/PoolManager.cpp]: Strike that... It's a vector, not an array. In which case I'd use:

`return &m_connections[m_activeConnectionIdx];`

I don't feel strongly about this. In my view, you either check for out-of-bound conditions before you use the index, or establish a proper try/catch construct and use the .at() accessor. Not both!

Comment (StefanOberhumer) [reply libpoolprotocols/PoolManager.cpp]: My personal favorite was `return &m_connections[m_activeConnectionIdx];` .
But I already implemented that a time ago while @AndreaLanfranchi  were searching bugs in the auto/connection. I think he replaced while searching the bug nearly all indexed array accesses (m_connections[]) with m_connections.at().
Only one m_connections[] access is currently in the PoolManager.cpp
I don't wanted to break the style so I also used m_connections.at().

Would it be possible that thread2 eliminates the element (after thread1 has checked the size ) and the then thread1 accesses an invalid element (eg if it was the last)?

Comment (AndreaLanfranchi) [reply libpoolprotocols/PoolManager.cpp]: m_activeConnectionIdx can be set by 2 threads (poolmgr and API) as well as the size of the vector although it's very rare for poolManager and API calls are supposed to be serialized (unless more than 1 API connection tries to issue commands ... which I expect unlikely).
For _extreme_ safety I'd go for a try/catch using .at() but in that case ... how do you think to handle the exception ? I mean ... which element would you return ?

Probably better to let getActiveConnection() throw and handle exception at higher level.
My 2 cents.

**Comment by jean-m-cyr** [comment]
Indeed, you have a race condition on m_connections, between pool manager and API server threads. Guess you'll need to add a lock around the uses of m_connections and m_activeConnectionIdx.

File: libapicore/ApiServer.cpp [hunk 2]
```diff
@@ -787,7 +787,7 @@ Json::Value ApiConnection::getMinerStat1()
 {
     auto runningTime = std::chrono::duration_cast<std::chrono::minutes>(
         steady_clock::now() - m_farm.farmLaunched());
-
+    auto connection = m_mgr.getActiveConnection();
```
Comment (jean-m-cyr) [inline libapicore/ApiServer.cpp]: This is bad. getActiveConnection returns a pointer to an object with no guarantee that the object will remain valid.

Comment (StefanOberhumer) [reply libapicore/ApiServer.cpp]: Will try to solve ...

**Comment by StefanOberhumer** [comment]
@jean-m-cyr , @AndreaLanfranchi  : Any comments are welcome...

File: libapicore/ApiServer.cpp (hunk 2 above)
Comment (StefanOberhumer) [reply libapicore/ApiServer.cpp]: ... and don't like locks ...

File: libpoolprotocols/PoolManager.cpp [hunk 3]
```diff
@@ -316,11 +316,17 @@ void PoolManager::setActiveConnection(unsigned int idx)
...
-const URI* PoolManager::getActiveConnection()
+URI PoolManager::getActiveConnectionCopy()
 {
-    if (m_connections.size() > m_activeConnectionIdx)
-        return &m_connections.at(m_activeConnectionIdx);
-    return nullptr;
+    URI r = URI(":0");
+    try
+    {
+       r = m_connections.at(m_activeConnectionIdx);
```
Comment (jean-m-cyr) [inline libpoolprotocols/PoolManager.cpp]: vectors are not thread safe. You could still end up with an invalid copy. For example: Say size = 5, and index = 3. The at() method only checks that index < size. What if at() is interrupted by a delete of index 2 by the ApiServer?

Comment (jean-m-cyr) [reply libpoolprotocols/PoolManager.cpp]: The copy is needed, but you also need to surround all accesses to m_connections and m_connection_index with a lock, to make it thread safe.

Comment (StefanOberhumer) [reply libpoolprotocols/PoolManager.cpp]: OK. I understand!
I've just moved the problem to an invisible (=std::vector)  .at() function where the same code [ if(vector.size()........] is done like I did it before.
... stay tuned ... next try will come ;-) ...

**Comment by StefanOberhumer** [comment]
** DON'T MERGE - DOES NOT WORK**

**Pull request merged**